### PR TITLE
OSDOCS-7776: Migration of "Adding a Public Ingress endpoint to a ROSA PrivateLink Cluster"

### PR DESCRIPTION
**Pull request by jneczypor**
[OSDOCS-7776](https://issues.redhat.com//browse/OSDOCS-7776): Migrate "Adding a Public Ingress endpoint to a ROSA PrivateLink Cluster" to ROSA product documentation

Version(s):
4.14+

Issue:
https://issues.redhat.com/browse/OSDOCS-7776

Link to docs preview:
https://64930--docspreview.netlify.app/openshift-rosa/latest/cloud_experts_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster

QE review:
- [ ] QE has approved this change.

Additional information:

### CODE REVIEW

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+image::arch-pl-ingress.png[Adding a public ingress endpoint to a ROSA cluster]
+
+For more information on adding public ingresses to a PrivateLink cluster  go link:https://developers.redhat.com/articles/2023/04/27/how-add-public-ingress-private-link-rosa-cluster[here].
+
+.Prerequisites
+
+* The link:https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html[AWS CLI]
+* The link:https://github.com/openshift/rosa/releases/tag/v1.0.8[ROSA CLI] (`rosa`)
+* The link:https://stedolan.github.io/jq/download/[`jq` command]
+* A link:./README.md[ROSA PrivateLink cluster]
```
Comment (EricPonvelle) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:36]: This link doesn't work.

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+== Getting Started
+
+=== Set your environment variables
+
+. Set the following environment variables, changing them to suit your cluster:
++
+[source,terminal]
+----
```
Comment (EricPonvelle) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:45]: This command block needs `$`

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+[NOTE]
+====
+If you followed the above instructions to create the ROSA PrivateLink cluster, you should already have a public subnet in your VPC and can skip to tagging the subnet.
+====
+
+. Obtain a private subnet ID from the cluster:
++
+[source,terminal]
+----
+   PRIVATE_SUBNET_ID=$(rosa describe cluster -c $ROSA_CLUSTER_NAME -o json \
```
Comment (EricPonvelle) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:66]: This command block needs `$`

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+      | jq -r '.aws.subnet_ids[0]')
+   echo $PRIVATE_SUBNET_ID
+----
+
+. Obtain the VPC ID from the subnet ID:
++
+[source,terminal]
+----
+    VPC_ID=$(aws ec2 describe-subnets --subnet-ids $PRIVATE_SUBNET_ID \
```
Comment (EricPonvelle) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:75]: This command block needs `$`

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+      --query 'Subnets[0].VpcId' --output text)
+    echo $VPC_ID
+----
+
+. Obtain the cluster tag from the subnet:
++
+[source,terminal]
+----
+   TAG=$(aws ec2 describe-subnets --subnet-ids $PRIVATE_SUBNET_ID \
```
Comment (EricPonvelle) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:84]: This command block needs `$`

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+      --query 'Subnets[0].Tags[?Value == `shared`]' | jq -r '.[0].Key')
+   echo $TAG
+----
+
+. Create a public subnet:
++
+[source,terminal]
+----
+   PUBLIC_SUBNET=`aws ec2 create-subnet --vpc-id $VPC_ID --cidr-block $PUBLIC_CIDR \
```
Comment (EricPonvelle) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:93]: This command block needs `$`

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+     --query 'Subnet.SubnetId' --output text`
+   echo $PUBLIC_SUBNET
+----
+
+. Tag the public subnet for the cluster:
++
+[source,terminal]
+----
+   aws ec2 create-tags --resources $PUBLIC_SUBNET \
```
Comment (EricPonvelle) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:102]: This command block needs `$`

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+. Create a TLS key pair for custom domain using certbot:
++
+[NOTE]
+====
+Skip this step if you already have a key pair.
+====
++
+[source,terminal]
+----
+   certbot certonly --manual \
```
Comment (EricPonvelle) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:118]: This command block needs `$`

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+. Create TLS secret for custom domain:
++
+[NOTE]
+====
+Use your own key pair paths if not using certbot.
+====
++
+[source,terminal]
+----
+   CERTS=/tmp/scratch/config/live/$DOMAIN
```
Comment (EricPonvelle) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:138]: This command block needs `$`

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+   oc new-project my-custom-route
+   oc create secret tls acme-tls --cert=$CERTS/fullchain.pem --key=$CERTS/privkey.pem
+----
+
+. Create a custom domain resource:
++
+[source,terminal]
+----
+   cat << EOF | oc apply -f -
```
Comment (EricPonvelle) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:147]: This command block needs `$`

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+       name: acme-tls
+       namespace: my-custom-route
+   EOF
+----
+
+. Wait for the domain to be ready:
++
+[source,terminal]
+----
+   watch oc get customdomains
```
Comment (EricPonvelle) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:164]: This command block needs `$`

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+----
+
+. Obtain the CLB name:
++
+[source,terminal]
+----
+   CDO_NAME=acme
```
Comment (EricPonvelle) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:171]: This command block needs `$`

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+
+. Create a CNAME in your DNS provider for *.<$DOMAIN> that points at the CLB NAME from the above command:
+
+=== Deploying a public application
+
+. Create a new project:
++
+[source,terminal]
+----
+   oc new-project my-public-app
```
Comment (EricPonvelle) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:184]: This command block needs `$`

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+----
+
+. Create a new application:
++
+[source,terminal]
+----
+   oc new-app --docker-image=docker.io/openshift/hello-openshift
```
Comment (EricPonvelle) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:191]: This command block needs `$`

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+----
+
+. Create a route for the application:
++
+[source,terminal]
+----
+   oc create route edge --service=hello-openshift hello-openshift-tls \
```
Comment (EricPonvelle) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:198]: This command block needs `$`

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+     --hostname hello.$DOMAIN
+----
+
+. Verify that you can access the application:
++
+[source,terminal]
+----
+   curl https://hello.$DOMAIN
```
Comment (EricPonvelle) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:206]: This command block needs `$`

**Review by EricPonvelle — COMMENTED** [review]
Small things.

**Comment by jneczypor** [comment]
Hi @michaelryanmcneill - Can you please review this PR? Thanks!

File: rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+$   echo $PUBLIC_SUBNET
+----
+
+. Tag the public subnet for the cluster:
++
+[source,terminal]
+----
+$   aws ec2 create-tags --resources $PUBLIC_SUBNET \
+      --tags Key=Name,Value=$ROSA_CLUSTER_NAME-public \
+      Key=$TAG,Value="shared" Key=kubernetes.io/role/elb,Value="true"
```
Comment (lihongan) [inline rosa_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:104]: The valid value of Key `kubernetes.io/role/elb` is empty or "1", please see https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.1/deploy/subnet_discovery/#public-subnets 
and other reference: https://docs.aws.amazon.com/eks/latest/userguide/network-load-balancing.html and https://docs.aws.amazon.com/eks/latest/userguide/alb-ingress.html

File: cloud_experts_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+. Use link:https://certbot.eff.org/[Certbot] to create a Transport Layer Security (TLS) key pair for the custom domain:
++
+[NOTE]
+====
+Skip this step if you already have a key pair.
+====
++
+[source,terminal]
+----
+$   certbot certonly --manual \
```
Comment (EricPonvelle) [inline cloud_experts_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:118]: All of these should be lined up under the `$`

**Review by EricPonvelle — COMMENTED** [review]
Small changes, but closer!

File: cloud_experts_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc (hunk 18 above)
Comment (jneczypor) [reply cloud_experts_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:118]: They are in VS Code but aren't appearing that way in GitHub for some reason.

File: cloud_experts_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+
+== Getting Started
+
+=== Setting your environment variables
+
+. Set the following environment variables, changing them to suit your cluster:
++
+[source,terminal]
+----
+$   export ROSA_CLUSTER_NAME=private-link
```
[the remark this answers is not in the document]
Comment (jneczypor) [reply cloud_experts_tutorials/rosa-mobb-adding-public-ingress-endpoint-private-link-cluster.adoc:46]: This is correct in VS Code.

**Pull request closed without merging**